juju/juju
# Add ingress rule support to the GCE provider

The GCE provider is updating to support the new ingress rules with source CIDRs.
The change looks bigger than is it because git decided the file move (ruleset.go from network -> provider/gce/google) was an add/remove.

The existing way the GCE provider managed port ranges was to explode them into individual ports. This is allowed for in the new implementation - we read in existing rules and collapse the port ranges. When we write out any changes, or create new rules port ranges are used.

We now create different firewalls corresponding to distinct CIDR sets to fit in with the GCE model. For the default 0.0.0.0/0 case, there is no change. As new CIDRs are added and removed, the firewall name will change to allow for distinct port ranges with their own CIDR set.

There's also a drive by tweak to the azure provider

QA
bootstrap GCE with older juju
expose an application
open ports 3100-3110
observe that the firewall rules are individual ports
upgrade juju
open ports 3120-3130
observe that the firewall is updated and now has port ranges 3100-3110,3120-3130
close ports 3100-3110
observe that the firewall is updated and now has port ranges 3120-3130
